compiler/codegen/cg-expr.cpp

```diff
@@ -4886,6 +4887,41 @@ DEFINE_PRIM(GPU_BLOCKDIM_Z)  { ret = codegenCallToPtxTgtIntrinsic("llvm.nvvm.rea
 DEFINE_PRIM(GPU_GRIDDIM_X)   { ret = codegenCallToPtxTgtIntrinsic("llvm.nvvm.read.ptx.sreg.nctaid.x"); }
 DEFINE_PRIM(GPU_GRIDDIM_Y)   { ret = codegenCallToPtxTgtIntrinsic("llvm.nvvm.read.ptx.sreg.nctaid.y"); }
 DEFINE_PRIM(GPU_GRIDDIM_Z)   { ret = codegenCallToPtxTgtIntrinsic("llvm.nvvm.read.ptx.sreg.nctaid.z"); }
+
+DEFINE_PRIM(GPU_ALLOC_SHARED) {
+#ifdef HAVE_LLVM
+  int bytesToAlloc = toVarSymbol(toSymExpr(call->get(1))->symbol())->immediate->int_value();
+
+  GenInfo* info = gGenInfo;
+  llvm::ArrayType* arrayTy = llvm::ArrayType::get(llvm::IntegerType::get(gGenInfo->llvmContext, 8), bytesToAlloc);
+  // Allocate global variable in "shared memory space" (3)
+  llvm::GlobalVariable* glob = new llvm::GlobalVariable(
+    *info->module, arrayTy, false, llvm::GlobalValue::InternalLinkage,
+    llvm::Constant::getNullValue(arrayTy),
+    "gpuSharedMemory", nullptr, llvm::GlobalValue::NotThreadLocal, 3, false); 
```

Comment:
Honestly, I ended up using this because it's the default value for the GlobalVariable class and I needed to explicitly supply it to get to the next parameter where I set the address space (3 for shared memory).

Here's what I can find in the LLVM documentation (https://llvm.org/docs/LangRef.html#tls-model): "A variable may be defined as thread_local, which means that it will not be shared by threads (each thread will have a separated copy of the variable)." So I think your understanding is right.

I'm not sure if this is the right model for us or not, or if this particular parameter is meaningful in the context of the shared-memory address space \ PTX code generation.

I suppose if you had a `cobegin` that launched two copies of the kernel you wouldn't want them sharing the same memory.

Now I'm wondering about how this works in the context of multiple, parallel, kernel launches and multiple GPUs as well.  Sounds like some ideas for interesting test cases.

Reply:
> Now I'm wondering about how this works in the context of multiple, parallel, kernel launches and multiple GPUs as well. Sounds like some ideas for interesting test cases.

Hmm... Interesting, but I don't anticipate any problems there. Obviously my understanding here is very weak.